### PR TITLE
tree-sitter: 0.18.2 -> 0.19.3

### PR DESCRIPTION
**Pull request by jlesquembre**
###### Motivation for this change
Update tree-sitter

Update the grammars and add `allGrammars` derivation.
`nix-build "<nixpkgs>" -A tree-sitter.allGrammars` will create a
derivation with symlinks to all the grammars

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/parsing/tree-sitter/default.nix [hunk 1]
```diff
@@ -32,35 +44,45 @@ let
...
+        { tree-sitter-tsx = grammars'.tree-sitter-typescript // { location = "tsx"; }; };
+    in
+    lib.mapAttrs change grammars;
+
+  allGrammars =
+    linkFarm "grammars"
+      (lib.attrsets.mapAttrsToList
+        (name: drv:
+          {
+            name = (lib.strings.removePrefix "tree-sitter-" name) + ".so";
```
Comment (SuperSandro2000) [inline pkgs/development/tools/parsing/tree-sitter/default.nix:79]: That breaks on darwin doesn't it?

File: pkgs/development/tools/parsing/tree-sitter/default.nix [hunk 2]
```diff
@@ -101,6 +123,7 @@ in rustPlatform.buildRustPackage {
     };
     inherit grammars;
     inherit builtGrammars;
+    inherit allGrammars;
```
Comment (SuperSandro2000) [inline pkgs/development/tools/parsing/tree-sitter/default.nix:126]:
```suggestion
    inherit grammars builtGrammars allGrammars;
```

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 115025` at 81d1cb77 run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li><a href="https://gist.github.com/cd804cb02a43c9ddfa84ef018e008657">tree-sitter</a></li></li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: unused-argument

Unused argument: `fetchurl`.
Near pkgs/development/tools/parsing/tree-sitter/default.nix:5:3:
```
  |
5 | , fetchurl
  |   ^
```
</li>
  </ul>
</details>

File: pkgs/development/tools/parsing/tree-sitter/default.nix (hunk 1 above)
Comment (jlesquembre) [reply pkgs/development/tools/parsing/tree-sitter/default.nix]: Right, I think is doesn't really matter, but now for darwin the extensions will be `.dylib`. I don't have a mac to test it

File: pkgs/development/tools/parsing/tree-sitter/default.nix (hunk 2 above)
Comment (jlesquembre) [reply pkgs/development/tools/parsing/tree-sitter/default.nix]: Thanks, updated

**Comment by jlesquembre** [comment]
Related PR: #115071

File: pkgs/development/tools/parsing/tree-sitter/default.nix [hunk 3]
```diff
@@ -32,35 +44,48 @@ let
...
+        { tree-sitter-ocaml = grammars'.tree-sitter-ocaml // { location = "ocaml"; }; } //
+        { tree-sitter-ocaml-interface = grammars'.tree-sitter-ocaml // { location = "interface"; }; } //
+        { tree-sitter-typescript = grammars'.tree-sitter-typescript // { location = "typescript"; }; } //
+        { tree-sitter-tsx = grammars'.tree-sitter-typescript // { location = "tsx"; }; };
+    in
+    lib.mapAttrs change grammars;
+
+  allGrammars =
+    let
+      ext = if stdenv.targetPlatform.isMacOS then ".dylib" else ".so";
```
Comment (teto) [inline pkgs/development/tools/parsing/tree-sitter/default.nix:76]: you can use `stdenv.hostPlatform.extensions.sharedLibrary` isnteaad

Comment (jlesquembre) [reply pkgs/development/tools/parsing/tree-sitter/default.nix]: :+1: done

**Comment by teto** [comment]
0.19.2 is out :s

**Comment by jlesquembre** [comment]
@ teto Updated to 0.19.2. Let's see if I have to update it again to 0.19.3...

I removed the `allGrammars` attribute I added for #115071. Instead, I created #115445, I think that's a better way to manage the grammars

**Comment by teto** [comment]
> Let's see if I have to update it again to 0.19.3...

Sry but looking at the treesitter tracker, seems like there is a 0.19.3 upcoming to address 0.19.2 issues :/

**Comment by jlesquembre** [comment]
> >  Let's see if I have to update it again to 0.19.3...

> Sry but looking at the treesitter tracker, seems like there is a 0.19.3 upcoming to address 0.19.2 issues :/

That was expected, at least every update takes me less time, I'm becoming an expert updating the tree-sitter grammars :laughing:

**Review by teto — APPROVED** [review]

**Comment by teto** [comment]
builds locally with my config (no nixpkgs-review)

**Comment by teto** [comment]
Some of the grammars dont load in neovim master (https://github.com/neovim/neovim/pull/14076 with 0.19.3 support). I am patching neovim to improve the error messsage.

**Comment by teto** [comment]
some grammars are not ready for 0.19 (lua/nix) apparently so be careful. Hopefully they will be updated shortly.

**Pull request merged**